src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -5463,9 +5481,38 @@ private static BoundExpression SkipReferenceConversions(BoundExpression possibly
             Join(ref this.State, ref leftState);
             TypeWithState resultType = TypeWithState.Create(targetType.Type, rightResult.State);
             SetResultType(node, resultType);
+
+            if (adjustedNodeType != (object?)node.Type)
+            {
+                Debug.Assert(adjustedNodeType is not null);
+                ForceDynamicResultType(node, adjustedNodeType);
+            }
+
             return null;
         }
 
+        private void ForceDynamicResultType(BoundExpression node, TypeSymbol sourceType)
```

Comment:
Although the method is simple (it's easy to see what it does), a comment with context will be helpful for future readers that wonder why scenarios need to do this.